INSTALL.md

```diff
@@ -0,0 +1,94 @@
+### Run-Time Compatibility
+
+Functions are provided to check the library configuration and solid harmonics orderings at runtime:
+
+Note: As of v2.8.0 (libtool-based), the configuration_accessor() function will return `(nyi)` by default.
+  Packagers are encouraged to patch a generated configuration string into file `configuration.cc` to
+  imitate future cmake-based behavior. See sample patch below.
+
+```
+libint2::initialize();
+printf("SHGShell: %d\n", libint2::solid_harmonics_ordering());
+libint2::set_solid_harmonics_ordering(libint2::SHGShellOrdering_Gaussian);
+printf("SHGShell: %d\n", libint2::solid_harmonics_ordering());
+printf("Configuration: %s\n", libint2::configuration_accessor().c_str());
+libint2::finalize();
+```
+```
+SHGShell: 1
+SHGShell: 2
+Configuration: eri_c4_d0_l2;eri_c4_d0_l3;ss;...
+```
+
+For the C library, a similar function is available:
+
+```
+printf("CMake Configuration (C)  : %s\n", configuration_accessor());
+```
+```
+CMake Configuration (C)  : eri_c4_d0_l2;eri_c4_d0_l3;ss;...
+```
+
+If you have a built libint2 library whose history you don't know, a command like this on Linux can provide the same information:
+
+```
+strings -n80 /a/random/L2/lying/around/libint2.so
+```
+```
+eri_c2_d0_l2;eri_c2_d0_l3;eri_c2_d1_l2;eri_c3_d0_l2;eri_c3_d0_l3;eri_c3_d1_l2;eri_c4_d0_l2;eri_c4_d1_l2;impure_sh;onebody_d0_l2;onebody_d0_l3;onebody_d1_l2;ss
+```
+
+A patch like the following is suitable for an export tarball generated from the next following.
+[See guide](#configuration-codes) for decoding the configuration components.
+
+```
+--- src/configuration.cc.cmake.in   2023-09-05 09:13:50.000000000 -0400
++++ src/configuration.cc.cmake.in_basic 2023-09-05 23:41:00.444396591 -0400
+@@ -24,6 +24,6 @@
+    @return the semicolon-separated strings from CMake components */
+ const char * configuration_accessor() {
+     //return "@Libint2_CONFIG_COMPONENTS@";
+-    return "(nyi)";
++    return "eri_c2_d0_l2;eri_c2_d0_l3;eri_c2_d0_l4;eri_c2_d0_l5;eri_c2_d0_l6;eri_c2_d1_l2;eri_c2_d1_l3;eri_c2_d1_l4;eri_c2_d1_l5;eri_c3_d0_l2;eri_c3_d0_l3;eri_c3_d0_l4;eri_c3_d0_l5;eri_c3_d0_l6;eri_c3_d1_l2;eri_c3_d1_l3;eri_c3_d1_l4;eri_c3_d1_l5;eri_c4_d0_l2;eri_c4_d0_l3;eri_c4_d0_l4;eri_c4_d0_l5;eri_c4_d1_l2;eri_c4_d1_l3;eri_c4_d1_l4;g12_d0_l2;g12_d0_l3;g12_d0_l4;g12_d1_l2;g12_d1_l3;g12_d1_l4;impure_sh;onebody_d0_l2;onebody_d0_l3;onebody_d0_l4;onebody_d0_l5;onebody_d0_l6;onebody_d1_l2;onebody_d1_l3;onebody_d1_l4;onebody_d1_l5;onebody_d2_l2;onebody_d2_l3;onebody_d2_l4;ss";
```

Comment:
I am slightly concerned about the lack of support for different angular momenta on different centers in component notation ... how about instead of `eri_c3_d1_l5` and `eri_c2_d1_l5` we say `eri_hhH_d1` and `eri_HH_d1`, with `h` and `H` denotting Cartesian and solid harmonic l=5 Gaussians? And instead of `impure_sh` we would make `eri_hhh_d1` and ``eri_hh_d1` available.

It is likely Psi does not use this currently, but generally this is useful since fitting basis should include higher l than the corresponding orbital basis, so it makes sense to have "fitting" centers to support higher l than the rest...

Reply:
Is the mixed AM on different centers specifiable through the build system now? (I suppose I've never specified AM for eri2 and eri3 differently.) I'm not quite following how `h` and `H` relate to generator input parameters. If `h=4` and `H=5`, would the component be `eri_hhH445_d1`? Otherwise wouldn't `h=4` and `H=4` with pure-sh=T/F look the same? 

Having two input variables (eri2 max AM and eri3 max AM)  can generate a lot more components when filling in all the lesser accessible components (e.g., 445 means 444, 335, 225, etc. all accessible and might be requested by a L2 consumer), but in a pinch I can outsource it to python and not rely on CMake math. 

Would 4-center remain unchanged, `eri_c4_d1_l5` or `eri_d1_l5`, or should it switch to the new form, like `eri_hhhh4444_d1`?

Reply:
> yes, this is doable, but none of the public releases use this. Note use of lmax in https://github.com/evaleev/libint/blob/master/src/bin/libint/build_libint.cc#L1145 vs. lmax_default in https://github.com/evaleev/libint/blob/master/src/bin/libint/build_libint.cc#L1146 ... so configuring with --with-max-am=6 --with-eri3-max-am=7 would allow computing (k|ii) integrals ...

Aha, good to know. Down the road in #259, this is going to require some changes, as I presently assume the generic max-am is simply the largest of eri-max-am, eri3-max-am, eri2-max-am, not an independent information source.

> we indeed would need to provide means of "comparison" for component strings, e.g. libint2::supports("eri_ffG") ... doing this in 2 places would be a total waste of effort, might be worth building a c++ helper code and calling it from cmake to avoid logic duplication

Ok, I see what you mean now about hhH turning into `ffG` w/distinguishable lowercase/caps rather than `334`. I'll test, but I'm pretty sure CMake component handling is case sensitive, so that'll be fine. Letters are nice for humans but a bit harder for computers to compare (though it just needs a translation string and a L=7 convention).

The c++ comparison helper code called from cmake I don't think is a great strategy, if I'm understanding it correctly. The key step is `libint2-config.cmake` accessed at `find_package(Libint2)` time. That's a fairly dumb code that works on strings and the presence/absence of files. I've never seen external functions called from it. And if there's cross-compiling happening, the helper code might not even run at detection time on detection hardware. So I think the long enumeration of accessible strings is sadly the best cmake-wise. I'd probably switch from the string being cmake-computed to it being Python-computed, adding a req'd Py to the first line `build_libint` https://github.com/loriab/libint/blob/new-cmake-2023-take2-b/INSTALL.md#prerequisites .

For background, my existing plan for configuration access was the following:
* _at build-against time_ consumer codes would use cmake components to find a suitable library (e.g., `find_package(Libint2 COMPONENTS eri_c4_d2_l6 ss)` (old notation).
* _at qc runtime_ consumer codes would call `libint2::configuration_accessor()` and do whatever they want with the big cmake-generated (in future) string. For instance, psi promptly turns it into a dictionary, https://github.com/psi4/psi4/blob/c6c6d7ec3509132e9bd3800af3ab87e10ed7184b/psi4/driver/p4util/util.py#L342 .

To address your `libint2::supports` proposal, I could have the `configuration_accessor` take an optional string argument, say `eri_ffG`. Then the fn splits at `;` and returns found or not. It's a lot dumber than the comparison fn you suggested, since the combinatorics are pre-generated, but the effect to the user is the same.

> I would prefer making format uniform for all cases, e.g. eri_hhhh_d1. What do you think?

Sure, I like uniform, too. I'll sketch out some input parameters to cmake component scenarios this afternoon to make sure I understand the conversion. How do you feel about retaining the explicitly enumerated component string for reasons described above?

Reply:
`eri_ffg_d0` (all lowercase; mixed AM) is a thing, right? That is, differing AM on the unpaired index is possible w/o `--enable-eri3-pure-sh=yes`. So the format string is `eri_hhL_dD` (always when eri3 enabled) and `eri_hhl_dD` (unless `--enable-eri3-pure-sh`)?

Reply:
> test, but I'm pretty sure CMake component handling is case sensitive, so that'll be fine. Letters are nice for humans but a bit harder for computers to compare (though it just needs a translation string and a L=7 convention). EDIT: Now that I think on it, if components run AM together `<h><h><H>` rather than label them `l<L>`, the vars _must_ be letters, not ints.

the reason I chose letters instead of digits is they scale to much higher angular momenta before requiring separators (`n` ~ l=10, so `nnO` works whereas 101011 does not) and we can encode both cartesian/solid harmonic differences compactly (`k` ~ Cartesian l=7, `K` ~ solid harmonic l=7).
 
> The c++ comparison helper code called from cmake I don't think is a great strategy, if I'm understanding it correctly. The key step is `libint2-config.cmake` accessed at `find_package(Libint2)` time. That's a fairly dumb code that works on strings and the presence/absence of files. I've never seen external functions called from it. And if there's cross-compiling happening, the helper code might not even run at detection time on detection hardware. So I think the long enumeration of accessible strings is sadly the best cmake-wise. I'd probably switch from the string being cmake-computed to it being Python-computed, adding a req'd Py to the first line `build_libint` https://github.com/loriab/libint/blob/new-cmake-2023-take2-b/INSTALL.md#prerequisites .

Yea, I agree that crosscompiling would throw in a wrench ... it would be awesome to have single definition of this logic, but it may not be practical.

I agree, cmake is the worst choice for writing such logic.

> > To address your `libint2::supports` proposal, I could have the `configuration_accessor` take an optional string argument, say `eri_ffG`. Then the fn splits at `;` and returns found or not. It's a lot dumber than the comparison fn you suggested, since the combinatorics are pre-generated, but the effect to the user is the same.

that would work, or if we are going full python for this code we can do the parse easily. The problem with enumerating all possibilities is that L^4 entries for large L can be quite large ... for heavy elements we will need full support for L=6 at least and with DF I can envision needing iiL ints (i.e. OBS max L = 6, DFBS max L = 8).

> Sure, I like uniform, too. I'll sketch out some input parameters to cmake component scenarios this afternoon to make sure I understand the conversion. How do you feel about retaining the explicitly enumerated component string for reasons described above?

Not too concerned, we can always change the logic later if it becomes unwieldy.

Reply:
> `eri_ffg_d0` (all lowercase; mixed AM) is a thing, right? That is, differing AM on the unpaired index is possible w/o `--enable-eri3-pure-sh=yes`. So the format string is `eri_hhL_dD` (always when eri3 enabled) and `eri_hhl_dD` (unless `--enable-eri3-pure-sh`)?

exactly. hence the use of letters instead of digits

Reply:
Whoops, I messed your comments from an hour ago. I'll read them next. I just pushed docs revised according to the component plan from this morning and a component list from the sample configuration.

Reply:
>> To address your libint2::supports proposal, I could have the configuration_accessor take an optional string argument, say eri_ffG. Then the fn splits at ; and returns found or not. It's a lot dumber than the comparison fn you suggested, since the combinatorics are pre-generated, but the effect to the user is the same.

> that would work, or if we are going full python for this code we can do the parse easily. The problem with enumerating all possibilities is that L^4 entries for large L can be quite large ... for heavy elements we will need full support for L=6 at least and with DF I can envision needing iiL ints (i.e. OBS max L = 6, DFBS max L = 8).

For configuration identification purposes, I was only thinking of e.g., `eri_iiii_d0`, `eri_hhhh_d0`, etc. since the different centers of a 4-center can't be specified independently in the buildsys (right?), so there's not the L^4 threat to the component list. So one wouldn't be calling `find_package(Libint2 COMPONENTS eri_ghgh_d0)` or `libint2::supports(eri_ghgh_d0)`. Do you foresee people needing to call `libint2::supports/configuration_acessor()` with that mixed/L^4 case?  I guess I figured consumers would be calling `configuration_accessor` once and collecting some AM/deriv limits for large-scale logic, not calling on all AM combinations. 

I suppose the only immediate decision for this PR is whether you want the `bool libint2::supports(std::string component) { /* pseudocode */ return component in configuration_accessor().split(";") }`.

---

I tentatively revised the `onebody_` component format for uniformity to `onebody_h_dD`. Not sure what to do on the g12 unless `g12_hhhh_dD`. I also never happened to put in components for multipole order; could be `multipole_h_dD` if you like.

Reply:
I went ahead and added `libint2::supports`. Known remaining issues include settling on onebody/g12/multipole? formats and updating the sample patch. I've got a couple CHANGES changes, too.